### PR TITLE
Annotations in Runtime

### PR DESCRIPTION
**Pull request by hichbra**
To be able to use Algorithms with Java reflection, I added 2 new Annotations in gs-algo:
- GSParameter : Use for methods in an algorithm which describe parameters. This annotation contains 1 value to define if the parameter is require. By default he is not.
- GSResult : The default output of an algorithm after execution.

Thus, it's possible to get all Algorithm (or only DynamicAlgorithm) with their interfaces and get annotated methods : 
```java
final List<Method> methods = new ArrayList<Method>();
Class<?> clazz = // Algorithm class
while (clazz != Object.class) { 
	final List<Method> allMethods = new ArrayList<Method>(Arrays.asList(clazz.getDeclaredMethods()));       
	for (final Method method : allMethods) {   	 
		if (method.isAnnotationPresent(GSParameter.class)) {
	        	methods.add(method);
	        }
	}
        clazz = clazz.getSuperclass();
}
```
-- Reviewer --
@gsavin 
@pigne 
@jordilaforge

### CODE REVIEW

**Comment by gsavin** [comment]
Sounds good!
Just splitting hair but is the `GS` prefix mandatory? No prefix would be ok I think, or an `Algorithm` prefix if you really need one?

**Comment by hichbra** [comment]
The prefix is not required, I added it to differentiate GraphStream annotations from others libraries. Furthermore, there are already a class Parameter (which is not used anywhere btw).

**Comment by hichbra** [comment]
I agree, I will remove the prefix and the old annotation. 

Regarding the DefineParameter features, I currently have an only trivial utilization of annotations, and for this purpose I did not need more elements than "require". 
param
But it will still be possible to add new attributes in "Parameter" annotation if it needed.

File: src/org/graphstream/algorithm/LongestPath.java [hunk 1]
```diff
@@ -129,6 +131,11 @@ public Path getLongestPath() {
         }
         return path;
     }
+    
+    @Result
+    public String defaultResult() {
```
Comment (jordilaforge) [inline src/org/graphstream/algorithm/LongestPath.java]: Why is the default result always a String? Shouldnt it be a graph or a collection of nodes?

Comment (hichbra) [reply src/org/graphstream/algorithm/LongestPath.java]: It's true that @Result annotation can be confusing. I created it with the meaning "Default result display". That's why you can find some algorithms with just a function "defaultMessage()" to inform users that the result was stored in attributes.

**Comment by hichbra** [comment]
That's right, these annotations were created to use reflection on the algorithms.
In practical terms, they not absolutely required, but recommended for some reasons. In my case, I need them to use and display algorithms in my application.
ex

**Comment by jordilaforge** [comment]
OK, that's fine but in general, Reflection should be used with care for a lot of reasons and often has the smell of bad design. In your case, I see that you are building an application to show off all available algorithms. Why not move the result method into the algorithm interface, therefore, every algorithm has to implement a result method (no reflection needed). The Problem will be the return value because some algos modify the graph and others return a path or a single node. I suggest you extend the Algorithm into different types of algorithm, for example, PathAlgorithm, SortAlgorithm and so on. Path algo would return a path, Sort algos a graph. The second problem will be the methods which set some attribute to change the behaviour of the algorithm. First, every algorithm should run without setting any attribute (default behaviour like my longestPath)  secondly the way an algorithms behaviour should be defined at the very beginning meaning in the constructor or like in graphstream in the init method. The simplest solution would be to add a varargs parameter to the init method in the algorithm interface which defines the behaviour of the algorithm. But might not be the best solution but just what crossed my mind while thinking about it. I am sure you can come up with a better solution. In general, it feels like in this PR you built a workaround for a design flaw in your application.  This is just my opinion you are the maintainer of this project. I do not want to insult anybody here....

**Comment by jordilaforge** [comment]
Another idea would be to put all settings for an algorithm inside a property file (config file) which is consumed be every algorithm at the beginning. Therefore you do not have to change the signature of the init method....

**Comment by hichbra** [comment]
Thank you for your interest!

> I suggest you extend the Algorithm into different types of algorithm, for example, PathAlgorithm, SortAlgorithm and so on. Path algo would return a path, Sort algos a graph.

I'm not sure about that. I am not the one who developed the existing system so I can't provide an answer for them, but the current architecture with Algorithm / DynamicAlgorithm allow a large degree of freedom to conception for the developer, in the functioning and the result. Also, trying to categorize algorithms may be difficult for the more exotic of them.

>the way an algorithms behaviour should be defined at the very beginning meaning in the constructor or like in graphstream in the init method.

Currently, all algorithms use their own method init with the required parameters. 
However, they also share a common method init defined in the interface which allows us have an overall approach of an algorithm, but only of course if we define the missing attributes with the specific methods.

> In general, it feels like in this PR you built a workaround for a design flaw in your application. 

As I said previously, GraphStream don't need annotation. But now, if anybody wants to use the reflection for any reason, I simply provide a solution for him. 
Annotations was planned for a long time, but they have never been used due to the complexity. I reduced the necessary information to the strict minimum needed, by taking as reference an ordinary application which could use the reflection.

Generally speaking, you made a constructive criticism of the architecture of gs-algo, and not of the annotations themselves. I simply apply these annotations in the actual system.
I'm not the main developer of this project, I think these problems deserve their own issues, so we can discuss with everyone concerned.

**Comment by jordilaforge** [comment]
Hey hichbra
Like I said I just wanted to share my opinion. Maybe we should have a discussion about the gs-algo in general. The problem for me working with graphstream is that the community seems to have very little interest in the library in general and that the maintainer of this repos do not manage it very often. My last PR in gs-algo took over half a year to be merged. That is why I abandoned graphstream for another graph library.

**Pull request merged**